File: mirdb-server/tests/accessibility.test.js
/**
 * Accessibility Compliance Tests
 * Owner: Scenario 12 - Accessibility Compliance
 *
 * Tests:
 * - Heading hierarchy audit
 * - Keyboard navigation
 * - ARIA labels presence
 * - Focus indicators visibility
 * - Color contrast ratios
 */

const fs = require('fs');
const path = require('path');

// Load HTML and CSS files
const htmlPath = path.join(__dirname, '../src/web/index.html');
const cssPath = path.join(__dirname, '../src/web/styles/main.css');
const darkModeCssPath = path.join(__dirname, '../src/web/styles/dark-mode.css');

let htmlContent;
let cssContent;
let darkModeCssContent;

beforeAll(() => {
    htmlContent = fs.readFileSync(htmlPath, 'utf8');
    cssContent = fs.readFileSync(cssPath, 'utf8');
    darkModeCssContent = fs.readFileSync(darkModeCssPath, 'utf8');
});

describe('Accessibility Compliance - Heading Hierarchy', () => {
    let document;

    beforeEach(() => {
        // Create a DOM from the HTML
        global.document.body.innerHTML = htmlContent.replace(/<script.*?<\/script>/gs, '');
        document = global.document;
    });

    test('should have exactly one h1 element', () => {
        const h1Elements = document.querySelectorAll('h1');
        expect(h1Elements.length).toBe(1);
    });

    test('should have proper h1-h6 hierarchy without skipped levels', () => {
        const headings = document.querySelectorAll('h1, h2, h3, h4, h5, h6');
        const headingLevels = Array.from(headings).map(h => parseInt(h.tagName[1]));

        // Verify no level is skipped
        for (let i = 1; i < headingLevels.length; i++) {
            const current = headingLevels[i];
            const previous = headingLevels[i - 1];
            // Allow same level, one level down, or going back up to any level
            const isValidSequence = current <= previous + 1;
            expect(isValidSequence).toBe(true);
        }
    });

    test('should start with h1 as the first heading', () => {
        const firstHeading = document.querySelector('h1, h2, h3, h4, h5, h6');
        expect(firstHeading).not.toBeNull();
        expect(firstHeading.tagName).toBe('H1');
    });

    test('should have descriptive heading content', () => {
        const headings = document.querySelectorAll('h1, h2, h3, h4, h5, h6');
        headings.forEach(heading => {
            const text = heading.textContent.trim();
            expect(text.length).toBeGreaterThan(0);
        });
    });

    test('should use semantic HTML elements', () => {
        // Check for header element
        expect(document.querySelector('header')).not.toBeNull();

        // Check for main element
        expect(document.querySelector('main')).not.toBeNull();

        // Check for footer element
        expect(document.querySelector('footer')).not.toBeNull();

        // Check for nav element
        expect(document.querySelector('nav')).not.toBeNull();

        // Check for section elements
        const sections = document.querySelectorAll('section');
        expect(sections.length).toBeGreaterThan(0);
    });

    test('should have lang attribute on html element', () => {
        // Parse full HTML to check html element
        const langMatch = htmlContent.match(/<html[^>]*lang=["']([^"']*)["']/);
        expect(langMatch).not.toBeNull();
        expect(langMatch[1]).toBe('en');
    });
});

describe('Accessibility Compliance - ARIA Labels', () => {
    let document;

    beforeEach(() => {
        global.document.body.innerHTML = htmlContent.replace(/<script.*?<\/script>/gs, '');
        document = global.document;
    });

    test('should have aria-label on logo element', () => {
        const logo = document.querySelector('.logo');
        expect(logo).not.toBeNull();
        expect(logo.getAttribute('aria-label')).toBeTruthy();
    });

    test('should have aria-label on navigation toggle button', () => {
        const navToggle = document.getElementById('nav-toggle');
        expect(navToggle).not.toBeNull();
        expect(navToggle.getAttribute('aria-label')).toBeTruthy();
    });

    test('should have aria-expanded on toggle buttons', () => {
        const navToggle = document.getElementById('nav-toggle');
        expect(navToggle).not.toBeNull();
        expect(navToggle.hasAttribute('aria-expanded')).toBe(true);
    });

    test('should have aria-controls linking toggle to menu', () => {
        const navToggle = document.getElementById('nav-toggle');
        const navMenu = document.getElementById('nav-menu');
        expect(navToggle).not.toBeNull();
        expect(navMenu).not.toBeNull();
        expect(navToggle.getAttribute('aria-controls')).toBe('nav-menu');
    });

    test('should have aria-label on navigation elements', () => {
        const mainNav = document.querySelector('nav[aria-label="Main navigation"]');
        expect(mainNav).not.toBeNull();

        const footerNav = document.querySelector('nav[aria-label="Documentation links"]');
        expect(footerNav).not.toBeNull();
    });

    test('should have aria-labelledby on sections with titles', () => {
        const quickStartSection = document.getElementById('quick-start');
        expect(quickStartSection).not.toBeNull();
        expect(quickStartSection.getAttribute('aria-labelledby')).toBe('quick-start-title');
    });

    test('should have aria-hidden on decorative SVG elements', () => {
        const decorativeSvg = document.querySelector('.logo svg');
        expect(decorativeSvg).not.toBeNull();
        expect(decorativeSvg.getAttribute('aria-hidden')).toBe('true');
    });

    test('should have aria-label on all copy buttons', () => {
        const copyButtons = document.querySelectorAll('.copy-btn');
        copyButtons.forEach(button => {
            expect(button.getAttribute('aria-label')).toBeTruthy();
        });
    });

    test('should have type attribute on all button elements', () => {
        const buttons = document.querySelectorAll('button');
        buttons.forEach(button => {
            expect(button.hasAttribute('type')).toBe(true);
        });
    });
});

describe('Accessibility Compliance - Keyboard Navigation', () => {
    let document;

    beforeEach(() => {
        global.document.body.innerHTML = htmlContent.replace(/<script.*?<\/script>/gs, '');
        document = global.document;
    });

    test('should have all interactive elements focusable', () => {
        const interactiveElements = document.querySelectorAll('a, button, input, select, textarea, [tabindex]');
        interactiveElements.forEach(element => {
            const tabindex = element.getAttribute('tabindex');
            // Should be focusable (no negative tabindex unless specifically hidden)
            if (tabindex !== null) {
                expect(parseInt(tabindex)).toBeGreaterThanOrEqual(-1);
            }
        });
    });

    test('should have navigation links as anchor elements', () => {
        const navLinks = document.querySelectorAll('.nav-link');
        navLinks.forEach(link => {
            expect(link.tagName).toBe('A');
            expect(link.hasAttribute('href')).toBe(true);
        });
    });

    test('should have buttons with proper button role', () => {
        const buttons = document.querySelectorAll('button');
        buttons.forEach(button => {
            // Buttons should be actual button elements (implicit role)
            expect(button.tagName).toBe('BUTTON');
        });
    });

    test('should have skip link or logical focus order', () => {
        // Check for skip link or verify logical document order
        const firstFocusable = document.querySelector('a, button, input, [tabindex="0"]');
        expect(firstFocusable).not.toBeNull();
    });

    test('navigation links should have visible text content', () => {
        const navLinks = document.querySelectorAll('.nav-link');
        navLinks.forEach(link => {
            const text = link.textContent.trim();
            expect(text.length).toBeGreaterThan(0);
        });
    });
});

describe('Accessibility Compliance - Focus Indicators', () => {
    test('should have focus styles for buttons', () => {
        // Check CSS for focus styles on buttons
        expect(cssContent).toMatch(/\.btn:focus[^{]*\{[^}]*outline/);
    });

    test('should have focus styles for copy buttons', () => {
        expect(cssContent).toMatch(/\.copy-btn:focus[^{]*\{[^}]*outline/);
    });

    test('should have focus styles for nav links', () => {
        // Nav links should have focus indication
        expect(cssContent).toMatch(/\.nav-link:hover|\.nav-link\.active|\.nav-link:focus/);
    });

    test('should have focus styles for footer links', () => {
        expect(cssContent).toMatch(/\.footer-link:focus[^{]*\{[^}]*outline/);
    });

    test('should have focus styles for modal close button', () => {
        expect(cssContent).toMatch(/\.modal-close:focus[^{]*\{[^}]*outline/);
    });

    test('should have focus styles for search input', () => {
        expect(cssContent).toMatch(/\.key-search-input:focus[^{]*\{/);
    });

    test('should have visible outline-offset for better visibility', () => {
        // Verify outline-offset is used for accessibility
        expect(cssContent).toMatch(/outline-offset:\s*\d+px/);
    });

    test('should use rust-primary color for focus indicators', () => {
        // Focus should use the brand color for consistency
        expect(cssContent).toMatch(/outline.*var\(--color-rust-primary\)/);
    });
});

describe('Accessibility Compliance - Color Contrast', () => {
    /**
     * Helper to convert hex color to RGB
     */
    function hexToRgb(hex) {
        const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
        return result ? {
            r: parseInt(result[1], 16),
            g: parseInt(result[2], 16),
            b: parseInt(result[3], 16)
        } : null;
    }

    /**
     * Calculate relative luminance per WCAG 2.1
     */
    function getLuminance(rgb) {
        const [r, g, b] = [rgb.r, rgb.g, rgb.b].map(v => {
            v = v / 255;
            return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
        });
        return 0.2126 * r + 0.7152 * g + 0.0722 * b;
    }

    /**
     * Calculate contrast ratio per WCAG 2.1
     */
    function getContrastRatio(color1, color2) {
        const lum1 = getLuminance(color1);
        const lum2 = getLuminance(color2);
        const lighter = Math.max(lum1, lum2);
        const darker = Math.min(lum1, lum2);
        return (lighter + 0.05) / (darker + 0.05);
    }

    test('should have primary text color with minimum 4.5:1 contrast', () => {
        // Primary text (#212121) on light background (#FAFAFA)
        const textColor = hexToRgb('#212121');
        const bgColor = hexToRgb('#FAFAFA');
        const ratio = getContrastRatio(textColor, bgColor);
        expect(ratio).toBeGreaterThanOrEqual(4.5);
    });

    test('should have secondary text color with minimum 4.5:1 contrast', () => {
        // Secondary text (#616161) on light background (#FAFAFA)
        const textColor = hexToRgb('#616161');
        const bgColor = hexToRgb('#FAFAFA');
        const ratio = getContrastRatio(textColor, bgColor);
        expect(ratio).toBeGreaterThanOrEqual(4.5);
    });

    test('should have header text with minimum 4.5:1 contrast', () => {
        // Light text (#FAFAFA) on dark header (#1A1A1A)
        const textColor = hexToRgb('#FAFAFA');
        const bgColor = hexToRgb('#1A1A1A');
        const ratio = getContrastRatio(textColor, bgColor);
        expect(ratio).toBeGreaterThanOrEqual(4.5);
    });

    test('should have rust primary color with minimum 4.5:1 contrast on white', () => {
        // Rust primary (#E65100) on white (#FFFFFF)
        const textColor = hexToRgb('#E65100');
        const bgColor = hexToRgb('#FFFFFF');
        const ratio = getContrastRatio(textColor, bgColor);
        // Note: This might not meet 4.5:1, typically used for large text (3:1)
        expect(ratio).toBeGreaterThanOrEqual(3.0);
    });

    test('should define accessible muted text color', () => {
        // Check that muted text is defined with sufficient contrast
        // Muted text (#808080) in dark mode context
        expect(cssContent).toMatch(/--color-text-muted:\s*#[0-9A-Fa-f]{6}/);
    });

    test('should have dark mode colors meeting contrast requirements', () => {
        // Dark mode primary text (#E0E0E0) on dark background (#121212)
        const textColor = hexToRgb('#E0E0E0');
        const bgColor = hexToRgb('#121212');
        const ratio = getContrastRatio(textColor, bgColor);
        expect(ratio).toBeGreaterThanOrEqual(4.5);
    });

    test('should have code block text with sufficient contrast', () => {
        // Code text (#f8f8f2) on code background (#2d2d2d)
        const textColor = hexToRgb('#f8f8f2');
        const bgColor = hexToRgb('#2d2d2d');
        const ratio = getContrastRatio(textColor, bgColor);
        expect(ratio).toBeGreaterThanOrEqual(4.5);
    });

    test('should have success color with sufficient contrast', () => {
        // Success text (white) on success background (#1e7e34 - darker green for 4.5:1)
        const textColor = hexToRgb('#FFFFFF');
        const bgColor = hexToRgb('#1e7e34');
        const ratio = getContrastRatio(textColor, bgColor);
        expect(ratio).toBeGreaterThanOrEqual(4.5);
    });
});

describe('Accessibility Compliance - Form Accessibility', () => {
    let document;

    beforeEach(() => {
        global.document.body.innerHTML = htmlContent.replace(/<script.*?<\/script>/gs, '');
        document = global.document;
    });

    test('should have placeholder text for search input', () => {
        // Verify HTML structure supports accessible forms
        // The key-search-input should have placeholder
        expect(cssContent).toMatch(/\.key-search-input::placeholder/);
    });

    test('should have proper input styling for visibility', () => {
        // Check input has visible borders
        expect(cssContent).toMatch(/\.key-search-input[^{]*\{[^}]*border/);
    });
});

describe('Accessibility Compliance - Screen Reader Support', () => {
    let document;

    beforeEach(() => {
        global.document.body.innerHTML = htmlContent.replace(/<script.*?<\/script>/gs, '');
        document = global.document;
    });

    test('should have proper document title', () => {
        const titleMatch = htmlContent.match(/<title>([^<]+)<\/title>/);
        expect(titleMatch).not.toBeNull();
        expect(titleMatch[1].length).toBeGreaterThan(0);
    });

    test('should have meta charset defined', () => {
        expect(htmlContent).toMatch(/<meta\s+charset=["']UTF-8["']/i);
    });

    test('should have viewport meta tag', () => {
        expect(htmlContent).toMatch(/<meta\s+name=["']viewport["']/i);
    });

    test('should have live region for toast notifications in UI module', () => {
        // Read the ui.js file to verify toast has aria-live
        const uiPath = path.join(__dirname, '../src/web/scripts/ui.js');
        const uiContent = fs.readFileSync(uiPath, 'utf8');
        expect(uiContent).toMatch(/aria-live/);
        expect(uiContent).toMatch(/role.*alert/);
    });

    test('should have modal with proper ARIA attributes in UI module', () => {
        const uiPath = path.join(__dirname, '../src/web/scripts/ui.js');
        const uiContent = fs.readFileSync(uiPath, 'utf8');
        expect(uiContent).toMatch(/role.*dialog/);
        expect(uiContent).toMatch(/aria-modal.*true/);
    });
});

File: mirdb-server/tests/accessibility_test.rs
//! Accessibility Compliance tests.
//! Owner: Scenario 15 - Accessibility Compliance
//!
//! Tests:
//! 1. Verify keyboard navigation - all interactive elements reachable via Tab key
//! 2. Verify visible focus indicators on all interactive elements
//! 3. Verify semantic HTML structure - heading hierarchy and landmark regions
//! 4. Verify all images have alt text (logo and icons)
//!
//! WCAG 2.1 AA Requirements:
//! - 4.5:1 contrast ratio for normal text
//! - 3:1 contrast ratio for large text
//! - Keyboard accessible interactive elements
//! - Visible focus indicators

use axum::{body::Body, http::Request};
use tower::ServiceExt;

use mirdb::web::routes::create_router;

/// Helper function to get homepage HTML content
async fn get_homepage_html() -> String {
    let app = create_router();
    let response = app
        .oneshot(Request::builder().uri("/").body(Body::empty()).unwrap())
        .await
        .unwrap();

    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    String::from_utf8_lossy(&body).to_string()
}

/// Helper function to get CSS content
async fn get_css_content() -> String {
    let app = create_router();
    let response = app
        .oneshot(
            Request::builder()
                .uri("/static/style.css")
                .body(Body::empty())
                .unwrap(),
        )
        .await
        .unwrap();

    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    String::from_utf8_lossy(&body).to_string()
}

// ============================================================================
// Test Case 1: Keyboard Navigation - Tab through all interactive elements
// Expected: All buttons, links, and form fields reachable via Tab key
// ============================================================================

/// Test that all interactive elements are keyboard accessible
/// Interactive elements should have no tabindex=-1 (which would prevent keyboard access)
#[tokio::test]
async fn test_keyboard_navigation_all_interactive_elements_accessible() {
    let html = get_homepage_html().await;

    // Verify skip link exists for keyboard navigation
    assert!(
        html.contains("skip-link") && html.contains("Skip to main content"),
        "Page should have a skip link for keyboard navigation"
    );

    // Verify skip link targets main content
    assert!(
        html.contains("href=\"#main-content\""),
        "Skip link should target main content"
    );

    // Verify main content has id for skip link target
    assert!(
        html.contains("id=\"main-content\""),
        "Main content should have id for skip link target"
    );

    // Verify theme toggle button is accessible (no tabindex=-1)
    assert!(
        html.contains("id=\"theme-toggle\"") && html.contains("class=\"theme-toggle\""),
        "Theme toggle button should exist"
    );
    assert!(
        !html.contains("theme-toggle\" tabindex=\"-1\""),
        "Theme toggle button should be keyboard accessible"
    );

    // Verify all form inputs have proper id attributes for accessibility
    assert!(
        html.contains("id=\"set-key\"") || html.contains("name=\"key\""),
        "SET form should have accessible key input"
    );
    assert!(
        html.contains("id=\"set-value\"") || html.contains("name=\"value\""),
        "SET form should have accessible value input"
    );
    assert!(
        html.contains("id=\"get-key\""),
        "GET form should have accessible key input"
    );
    assert!(
        html.contains("id=\"delete-key\""),
        "DELETE form should have accessible key input"
    );

    // Verify navigation links are keyboard accessible
    assert!(
        html.contains("class=\"github-link\"") && html.contains("href=\"https://github.com"),
        "GitHub link should be keyboard accessible"
    );

    // Verify hero action buttons/links exist
    assert!(
        html.contains("class=\"btn") && html.contains("href=\"#"),
        "Hero action buttons should be keyboard accessible links"
    );
}

/// Test that all buttons have type attribute for keyboard accessibility
#[tokio::test]
async fn test_buttons_have_type_attribute() {
    let html = get_homepage_html().await;

    // Verify submit buttons exist in forms
    assert!(
        html.contains("type=\"submit\""),
        "Forms should have submit buttons with type attribute"
    );

    // Count submit buttons - should be at least 3 (SET, GET, DELETE)
    let submit_count = html.matches("type=\"submit\"").count();
    assert!(
        submit_count >= 3,
        "Should have at least 3 submit buttons for SET, GET, DELETE operations"
    );
}

// ============================================================================
// Test Case 2: Check Focus Indicators
// Expected: Visible focus ring on all interactive elements
// ============================================================================

/// Test that CSS includes focus styles for all interactive elements
#[tokio::test]
async fn test_focus_indicators_visible_in_css() {
    let css = get_css_content().await;

    // Verify global focus-visible styles exist
    assert!(
        css.contains(":focus-visible") || css.contains(":focus"),
        "CSS should include focus-visible or focus styles"
    );

    // Verify button focus styles
    assert!(
        css.contains(".btn:focus") || css.contains(".btn:focus-visible"),
        "CSS should include button focus styles"
    );

    // Verify input focus styles
    assert!(
        css.contains("input:focus") || css.contains("input:focus-visible"),
        "CSS should include input focus styles"
    );

    // Verify theme toggle focus styles
    assert!(
        css.contains(".theme-toggle:focus"),
        "CSS should include theme toggle focus styles"
    );

    // Verify GitHub link focus styles
    assert!(
        css.contains(".github-link:focus"),
        "CSS should include GitHub link focus styles"
    );

    // Verify focus uses outline property for visibility
    assert!(
        css.contains("outline:") || css.contains("outline-color"),
        "Focus styles should use outline for visibility"
    );

    // Verify focus outline is at least 2px for visibility
    assert!(
        css.contains("outline: 3px") || css.contains("outline: 2px"),
        "Focus outline should be at least 2px wide for visibility"
    );
}

/// Test that skip link has focus styles
#[tokio::test]
async fn test_skip_link_has_focus_styles() {
    let css = get_css_content().await;

    // Verify skip link class exists
    assert!(
        css.contains(".skip-link"),
        "CSS should include skip-link styles"
    );

    // Verify skip link focus styles
    assert!(
        css.contains(".skip-link:focus"),
        "Skip link should have focus styles"
    );
}

// ============================================================================
// Test Case 3: Run automated accessibility checker (semantic structure)
// Expected: No critical or serious accessibility violations
// ============================================================================

/// Test semantic HTML structure - proper heading hierarchy
#[tokio::test]
async fn test_semantic_html_heading_hierarchy() {
    let html = get_homepage_html().await;

    // Verify h1 exists (should be one main heading)
    let h1_count = html.matches("<h1").count();
    assert_eq!(h1_count, 1, "Page should have exactly one h1 heading");

    // Verify h2 exists (hero title)
    assert!(html.contains("<h2"), "Page should have h2 headings");

    // Verify h3 exists (section titles)
    assert!(html.contains("<h3"), "Page should have h3 section headings");

    // Verify h4 exists (subsection titles)
    assert!(html.contains("<h4"), "Page should have h4 subsection headings");

    // Verify headings follow proper order (h1 before h2 before h3)
    let h1_pos = html.find("<h1").expect("h1 should exist");
    let h2_pos = html.find("<h2").expect("h2 should exist");
    let h3_pos = html.find("<h3").expect("h3 should exist");
    assert!(
        h1_pos < h2_pos && h2_pos < h3_pos,
        "Headings should follow proper hierarchy order"
    );
}

/// Test semantic HTML structure - landmark regions
#[tokio::test]
async fn test_semantic_html_landmark_regions() {
    let html = get_homepage_html().await;

    // Verify header landmark
    assert!(
        html.contains("<header") && html.contains("role=\"banner\""),
        "Page should have header landmark with banner role"
    );

    // Verify main landmark
    assert!(
        html.contains("<main") && html.contains("role=\"main\""),
        "Page should have main landmark with main role"
    );

    // Verify footer landmark
    assert!(
        html.contains("<footer") && html.contains("role=\"contentinfo\""),
        "Page should have footer landmark with contentinfo role"
    );

    // Verify navigation landmark
    assert!(
        html.contains("<nav") && html.contains("role=\"navigation\""),
        "Page should have nav landmark with navigation role"
    );

    // Verify navigation has aria-label
    assert!(
        html.contains("aria-label=\"Main navigation\""),
        "Navigation should have aria-label"
    );
}

/// Test semantic HTML structure - sections have proper labeling
#[tokio::test]
async fn test_sections_have_proper_labeling() {
    let html = get_homepage_html().await;

    // Verify sections use aria-labelledby
    assert!(
        html.contains("aria-labelledby=\"hero-title\""),
        "Hero section should have aria-labelledby"
    );
    assert!(
        html.contains("aria-labelledby=\"features-title\""),
        "Features section should have aria-labelledby"
    );
    assert!(
        html.contains("aria-labelledby=\"dashboard-title\""),
        "Dashboard section should have aria-labelledby"
    );
    assert!(
        html.contains("aria-labelledby=\"try-title\""),
        "Try It Out section should have aria-labelledby"
    );
    assert!(
        html.contains("aria-labelledby=\"protocol-title\""),
        "Protocol section should have aria-labelledby"
    );
}

/// Test that forms have proper accessibility attributes
#[tokio::test]
async fn test_forms_have_proper_accessibility() {
    let html = get_homepage_html().await;

    // Verify forms have aria-labelledby
    assert!(
        html.contains("aria-labelledby=\"set-form-title\"")
            || html.contains("<form id=\"set-form\""),
        "SET form should have proper labeling"
    );
    assert!(
        html.contains("aria-labelledby=\"get-form-title\"")
            || html.contains("<form id=\"get-form\""),
        "GET form should have proper labeling"
    );
    assert!(
        html.contains("aria-labelledby=\"delete-form-title\"")
            || html.contains("<form id=\"delete-form\""),
        "DELETE form should have proper labeling"
    );

    // Verify inputs have labels (either label elements or aria-label)
    assert!(
        html.contains("<label") || html.contains("aria-label"),
        "Form inputs should have labels"
    );

    // Verify required inputs have aria-required
    assert!(
        html.contains("aria-required=\"true\"") || html.contains("required"),
        "Required inputs should indicate required status"
    );
}

/// Test result area has proper live region attributes
#[tokio::test]
async fn test_result_area_has_live_region() {
    let html = get_homepage_html().await;

    // Verify result div has aria-live for screen reader announcements
    assert!(
        html.contains("id=\"result\"") && html.contains("aria-live=\"polite\""),
        "Result area should have aria-live='polite' for screen reader announcements"
    );

    // Verify result div has role="status"
    assert!(
        html.contains("role=\"status\""),
        "Result area should have role='status'"
    );
}

// ============================================================================
// Test Case 4: Verify all images have alt text
// Expected: Logo and icons have descriptive alt attributes
// ============================================================================

/// Test that logo has accessible labeling
#[tokio::test]
async fn test_logo_has_accessible_labeling() {
    let html = get_homepage_html().await;

    // Verify logo icon has accessible label (role="img" with aria-label, or aria-hidden if decorative)
    let has_accessible_logo = html.contains("role=\"img\"") && html.contains("aria-label")
        || html.contains("aria-hidden=\"true\"");
    assert!(
        has_accessible_logo,
        "Logo icon should have accessible labeling (aria-label or aria-hidden)"
    );

    // Verify logo section exists
    assert!(
        html.contains("class=\"logo\""),
        "Logo section should exist"
    );

    // Verify MirDB text provides accessible name
    assert!(
        html.contains("<h1>MirDB</h1>"),
        "MirDB heading provides accessible name for logo"
    );
}

/// Test that decorative icons have aria-hidden
#[tokio::test]
async fn test_decorative_icons_have_aria_hidden() {
    let html = get_homepage_html().await;

    // Verify feature icons have aria-hidden (they are decorative)
    assert!(
        html.contains("class=\"feature-icon\"") && html.contains("aria-hidden=\"true\""),
        "Decorative feature icons should have aria-hidden='true'"
    );

    // Verify theme icon has aria-hidden (decorative)
    assert!(
        html.contains("class=\"theme-icon\"") && html.contains("aria-hidden=\"true\""),
        "Theme icon should have aria-hidden='true' as it is decorative"
    );
}

/// Test that theme toggle button has accessible label
#[tokio::test]
async fn test_theme_toggle_has_accessible_label() {
    let html = get_homepage_html().await;

    // Verify theme toggle has aria-label
    assert!(
        html.contains("id=\"theme-toggle\"") && html.contains("aria-label"),
        "Theme toggle button should have aria-label"
    );

    // Verify aria-label describes the action
    assert!(
        html.contains("Toggle") || html.contains("theme"),
        "Theme toggle aria-label should describe the toggle action"
    );
}

/// Test that external links have proper accessible attributes
#[tokio::test]
async fn test_external_links_accessibility() {
    let html = get_homepage_html().await;

    // Verify GitHub link has aria-label indicating external link
    assert!(
        html.contains("github-link") && html.contains("aria-label"),
        "GitHub link should have aria-label"
    );

    // Verify external link indicates it opens in new tab
    assert!(
        html.contains("opens in new tab") || html.contains("target=\"_blank\""),
        "External link should indicate it opens in new tab"
    );

    // Verify noopener for security
    assert!(
        html.contains("rel=\"noopener"),
        "External links should have rel='noopener' for security"
    );
}

// ============================================================================
// Additional accessibility tests
// ============================================================================

/// Test HTML lang attribute is set
#[tokio::test]
async fn test_html_has_lang_attribute() {
    let html = get_homepage_html().await;

    assert!(
        html.contains("lang=\"en\""),
        "HTML element should have lang attribute for accessibility"
    );
}

/// Test page has meta viewport for responsive design
#[tokio::test]
async fn test_page_has_viewport_meta() {
    let html = get_homepage_html().await;

    assert!(
        html.contains("viewport") && html.contains("width=device-width"),
        "Page should have viewport meta tag for responsive/accessible design"
    );
}

/// Test page has descriptive title
#[tokio::test]
async fn test_page_has_descriptive_title() {
    let html = get_homepage_html().await;

    assert!(
        html.contains("<title>") && html.contains("MirDB"),
        "Page should have descriptive title"
    );
}

/// Test sr-only class exists in CSS for screen reader content
#[tokio::test]
async fn test_sr_only_class_exists() {
    let css = get_css_content().await;

    // Verify sr-only class exists
    assert!(
        css.contains(".sr-only"),
        "CSS should include sr-only class for screen reader only content"
    );

    // Verify sr-only uses proper technique
    assert!(
        css.contains("clip") || css.contains("position: absolute"),
        "sr-only should use proper hiding technique"
    );
}

/// Test metrics have live region updates
#[tokio::test]
async fn test_metrics_have_live_regions() {
    let html = get_homepage_html().await;

    // Verify metric values have aria-live for updates
    assert!(
        html.contains("id=\"memory-value\"") && html.contains("aria-live"),
        "Memory metric should have aria-live for screen reader updates"
    );
    assert!(
        html.contains("id=\"disk-value\"") && html.contains("aria-live"),
        "Disk metric should have aria-live for screen reader updates"
    );
    assert!(
        html.contains("id=\"keys-value\"") && html.contains("aria-live"),
        "Keys metric should have aria-live for screen reader updates"
    );
    assert!(
        html.contains("id=\"connections-value\"") && html.contains("aria-live"),
        "Connections metric should have aria-live for screen reader updates"
    );
}
